mysql_tests/test_declarative.py

```diff
@@ -0,0 +1,305 @@
+
+
+async def test_table_args():
+    class Model(db.Model):
+        __tablename__ = "model1"
+
+    assert Model.__table__.implicit_returning
+
+    class Model(db.Model):
+        __tablename__ = "model2"
+
+        __table_args__ = dict(implicit_returning=False)
+
+    assert not Model.__table__.implicit_returning
+
+    class Model(db.Model):
+        __tablename__ = "model3"
+
+        __table_args__ = db.Column("new_col"), dict(implicit_returning=False)
+
+    assert not Model.__table__.implicit_returning
+    assert not hasattr(Model, "new_col")
+    assert not hasattr(Model.__table__.c, "nonexist")
+    assert hasattr(Model.__table__.c, "new_col")
+
+    class Model(db.Model):
+        __tablename__ = "model4"
+        __table_args__ = db.Column("col1"), db.Column("col2")
+
+        col3 = db.Column()
+
+    assert not hasattr(Model, "col1")
+    assert not hasattr(Model, "col2")
+    assert hasattr(Model, "col3")
+    assert hasattr(Model.__table__.c, "col1")
+    assert hasattr(Model.__table__.c, "col2")
+    assert hasattr(Model.__table__.c, "col3")
+
+    class Model(db.Model):
+        @db.declared_attr
+        def __tablename__(cls):
+            return "model5"
+
+    assert Model.__table__.name == "model5"
+
+
+async def test_inline_constraints_and_indexes(bind, engine):
+    u = await User.create(nickname="test")
+    us1 = await UserSetting.create(user_id=u.id, setting="skin", value="blue")
+
+    # PrimaryKeyConstraint
+    with pytest.raises(IntegrityError):
+        await UserSetting.create(id=us1.id, user_id=u.id, setting="key1", value="val1")
+
+    # ForeignKeyConstraint
+    with pytest.raises(IntegrityError):
+        await UserSetting.create(user_id=42, setting="key2", value="val2")
+
+    # UniqueConstraint
+    with pytest.raises(IntegrityError):
+        await UserSetting.create(
+            user_id=u.id, setting="skin", value="duplicate-setting"
+        )
+
+    # MySQL doesn't support CheckConstraint
+    # with pytest.raises(CheckViolationError):
+    #     await UserSetting.create(user_id=u.id, setting="key3", value="val3", col1=42)
+
+    # Index
+    status, result = await engine.status(
+        "SHOW INDEXES FROM `gino_user_settings` WHERE Key_name = 'col2_idx'"
+    )
+    assert status == 1
+
+
+async def test_join_t112(engine):
+    class Car(db.Model):
+        __tablename__ = "cars"
+
+        id = db.Column(db.BigInteger(), primary_key=True)
+
+    class Wheel(db.Model):
+        __tablename__ = "wheels"
+
+        id = db.Column(db.BigInteger(), primary_key=True)
+        car_id = db.Column(db.ForeignKey("cars.id"))
+
+    sql = (
+        "SELECT wheels.id, wheels.car_id, cars.id \nFROM wheels "
+        "INNER JOIN cars ON cars.id = wheels.car_id"
+    )
+
+    assert engine.compile(Wheel.join(Car).select())[0] == sql
+
+
+async def test_mixin():
+    class Tracked:
+        created = db.Column(db.DateTime(timezone=True))
+
+        @db.declared_attr
+        def unique_id(cls):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5587950)